### ISSUE TITLE
Perform additional checks when assigning callable to instance variable

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6534]
Pyright currently doesn't catch the following runtime error.

```python
from typing import Callable

class A:
    x: Callable[[], None]

def func() -> None:
    ...

a = A()

A.x = staticmethod(func)  # Assignment OK; no type violation
A.x()  # OK
a.x()  # OK

A.x = func  # Assignment OK; no type violation
A.x()  # OK
a.x()  # Runtime error: lambda takes 0 positional arguments but 1 was given

a.x = func  # Assignment OK; no type violation
a.x()  # OK
```

I don't know if this can be fixed without creating false positives. Needs investigation.

**Comment by erictraut** [comment]
The problem here is that in Python, it's commonplace for class variables to shadow instance variables. For some types, this is fine. For other types (such as callables), it's problematic because they are treated differently depending on whether they're accessed via the class or an instance of the class. Pyright doesn't know when the value is stored how it is intended to be used unless it is explicitly declared as a "pure class variable" or a "pure instance variable". In cases where it's both, there's no good way to know.

I'm going to close this for now, but I've noted that it's a problematic open issue that we should revisit in the formal typing spec.
